### PR TITLE
Add the option to specify a specific key id format that is generated …

### PR DESCRIPTION
**Pull request by broamski**
…on the backend

### CODE REVIEW

**Comment by jefferai** [comment]
I think if we're going to add this we need to not assume that everyone will always want the display name (since in most cases it's relatively useless). A better approach would be to use something a bit more templated, as in the SQL databases where you can provide a string with `{{username}}` and so.

Potential values could be `{{token_display_name}}`, `{{role_name}}`, and maybe some reasonable others.

**Comment by broamski** [comment]
I agree with your recommendation, so I implemented it!

**Comment by jefferai** [comment]
Looks good. I realize the SSH CA tests aren't the easiest to deal with but could you try adding a test for this?

**Comment by broamski** [comment]
Hi @jefferai - I added the requested test and an additional test that checks the enforcement of `allow_user_key_ids`.

File: builtin/logical/ssh/path_sign.go [hunk 1]
```diff
@@ -275,14 +274,20 @@ func (b *backend) calculateKeyId(data *framework.FieldData, req *logical.Request
...
 
-	if req.DisplayName != "" {
-		keyId = fmt.Sprintf("%s-%s", req.DisplayName, keyId)
+	if role.KeyIDFormat != "" {
+		keyIdFormat = role.KeyIDFormat
 	}
 
-	keyId = fmt.Sprintf("vault-%s", keyId)
+	keyId := substQuery(keyIdFormat, map[string]string{
+		"token_display_name": req.DisplayName,
```
Comment (vishalnayak) [inline builtin/logical/ssh/path_sign.go]: What happens when role's `KeyIDFormat` mentions `{{token_display_name}}` and `req.DisplayName` is empty? I think we'll end up having a `{{}}` substring in `keyId`. Given that the role specified it, it seems reasonable to have an `{{}}` block. Regardless, can we have a comment here explaining that?

File: builtin/logical/ssh/path_sign.go [hunk 2]
```diff
@@ -275,14 +274,20 @@ func (b *backend) calculateKeyId(data *framework.FieldData, req *logical.Request
...
+	}
 
-	if req.DisplayName != "" {
-		keyId = fmt.Sprintf("%s-%s", req.DisplayName, keyId)
+	if role.KeyIDFormat != "" {
+		keyIdFormat = role.KeyIDFormat
 	}
 
-	keyId = fmt.Sprintf("vault-%s", keyId)
+	keyId := substQuery(keyIdFormat, map[string]string{
```
Comment (vishalnayak) [inline builtin/logical/ssh/path_sign.go]: Nitpick: Can we replace `Id` by `ID`?

File: builtin/logical/ssh/path_sign.go (hunk 1 above)
Comment (broamski) [inline builtin/logical/ssh/path_sign.go]: If req.DisplayName is an empty string, the return value should be the same. Here's an example: https://play.golang.org/p/um-WzsD8RE

File: builtin/logical/ssh/path_sign.go (hunk 2 above)
Comment (broamski) [inline builtin/logical/ssh/path_sign.go]: Fixed.

**Pull request merged**